packages/babel-plugin-transform-parameters/src/rest.js

```diff
@@ -155,6 +155,16 @@ const memberExpressionOptimisationVisitor = {
     }
   },
 };
+
+function getParamsCount(node) {
+  let count = node.params.length;
+  // skip the first parameter if it is a TypeScript 'this parameter'
+  if (count > 0 && t.isIdentifier(node.params[0], { name: "this" })) {
```

Comment:
Should we be checking that there's a `typeAnnotation` attached to this?

Reply:
I don't think that it is needed: `this` is not a valid binding identifier, so it can't be anything else other than TypeScript.